src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -33,38 +41,73 @@ public ODataInnerError(Exception exception)
         {
             ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
-
             if (exception.InnerException != null)
             {
                 this.InnerError = new ODataInnerError(exception.InnerException);
             }
+
+            Properties = new Dictionary<string, ODataValue>();
+
+            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
+            ////       to stay compatible with Astoria.
+            Properties.Add(JsonConstants.ODataErrorInnerErrorMessageName, exception.Message.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorTypeNameName, exception.GetType().FullName.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorStackTraceName, exception.StackTrace.ToODataValue() ?? new ODataNullValue());
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with a dictionary of property names and corresponding ODataValues.
+        /// </summary>
+        /// <param name="properties">Dictionary of string keys with ODataValue as value. Key string indicates the property name where as the value of the property is encapsulated in ODataValue.</param>
+        /// <param name="nestedInnerErrorName">The name of the field for the nested inner error object inside inner error.</param>
+        /// <param name="nestedObject">Nested inner error object.</param>
+        public ODataInnerError(IDictionary<string, ODataValue> properties)
+        {
+            ExceptionUtils.CheckArgumentNotNull(properties, "properties");
+
+            Properties = properties;
+        }
+
+        /// <summary>
+        /// Creates an instance of ODataInnerError with an empty property bag.
+        /// </summary>
+        /// <returns>returns an instance of ODataInnerError with empty property bag.</returns>
+        public static ODataInnerError CreateEmptyInnerError()

```

Comment:
> ODataInnerError CreateEmptyInnerError() [](start = 21, length = 40)

Is this a common enough pattern that we need the method, rather than having the caller simply create a new ODataInnerError with an empty dictionary (i.e., the line below)?

Reply:
It will be commonly used in the error handling library but I suppose that it is not a must. However, I wanted to include this as the name clarifies that there will be no auto-populated/default properties.

---
In reply to: [321857649](https://github.com/OData/odata.net/pull/1523#discussion_r321857649) [](ancestors = 321857649)

Reply:
Then do we really need the constructor that takes an IDictionary<string,ODataValue>, or can the error handling library just use this method and then add the properties directly to the ODataInnerError.Properties dictionary?

---
In reply to: [321869037](https://github.com/OData/odata.net/pull/1523#discussion_r321869037) [](ancestors = 321869037,321857649)